Fix bug at ssh.py:process() - empty argv[0] Error (#2217)

goreil opened the pull request:
Reopening #2225, since the Coverage tests didn't seem to handle changing the target branch correctly.

This addresses #2217 , a bug which disallowed us to use `ssh.py:process()` with an empty argv[0]

# Bug description
`ssh.py.process()` uses a python script to call `os.execve` to create a new process. `os.execve` disallows an empty argv.

# Fix
 This commit replaces `os.execve` with  `ctypes.CDLL(None).execve` to call the C-Library function directly which allows an empty argv.

goreil committed:
Fix bug at ssh.py:process() - empty argv[0] Error

Before this, process.py relied on `os.execve` which disallows an empty
argv or argv[0] == "".
This commit replaces `os.execve` with  `ctypes.CDLL(None).execve` to
call the C-Library function directly which allows an empty argv.

goreil committed:
Add #2225 to stable changelog

goreil committed:
Better ctypes syntax

goreil committed:
Updata CHANGELOG.md

goreil committed:
Add error message if cytpes.execve fails.

goreil committed:
Merge branch 'dev' into stable

goreil committed:
Merge branch 'Gallopsled:stable' into stable

goreil committed:
Merge branch 'dev' into stable

goreil commented:
Hi, just wanted to ask if there are any changes you require me to make to the pull request or is it ready to merge?

Arusekk commented:
For now I can only tell I am 90% sure that `environb` will break under py2.  I should find some time for a deeper review soon.

goreil committed:
ssh.py: python2 compatibility for os.environb

goreil committed:
Merge branch 'stable' of github.com:Youheng-Lue/pwntools into stable

goreil commented:
> For now I can only tell I am 90% sure that `environb` will break under py2. I should find some time for a deeper review soon.

Wow, totally missed that. Thanks for the heads up! Fixed.

Arusekk committed:
Merge branch 'dev' into stable

peace-maker commented:
https://github.com/Gallopsled/pwntools/pull/2225#issuecomment-1635984358

The code checks the environment variable name not the value.

peace-maker commented on pwnlib/tubes/ssh.py:
```diff
@@ -1031,7 +1048,29 @@ def is_exe(path):
 %(func_src)s
 %(func_name)s(*%(func_args)r)
 
-os.execve(exe, argv, env)
```
Can we only use the ctypes workaround if `exe != argv[0]`?

peace-maker commented on pwnlib/tubes/ssh.py:
```diff
@@ -1031,7 +1048,29 @@ def is_exe(path):
+# os.execve(exe, argv, env)
+        
+# Transform envp from dict to list
+env_list = [key + b"=" + value for key, value in env.items()]
```
Please [add a check](https://github.com/python/cpython/blob/025995feadaeebeef5d808f2564f0fd65b704ea5/Modules/posixmodule.c#L6476) that `key` does not contain `=`.

peace-maker commented on pwnlib/tubes/ssh.py:
```diff
@@ -1031,7 +1048,29 @@ def is_exe(path):
+c_env = to_carray(env_list)
+
+# Call execve
+libc = ctypes.CDLL(None)
```
Do you have a reference of where `ctypes.CDLL` accepts `None` as `name` to get libc? Above we're using `ctypes.CDLL('libc.so.6')`.

goreil committed:
Add check that "=" not in misc.normalize_argv_env

This check checks prevents the use of "=" in the
key of an environment variable, which is generally
impossible.

goreil committed:
ssh.process: Seperate cases for empty argv[0]

This commit seperates the cases for an empty argv[0]
from normal cases.

peace-maker committed:
Merge branch 'dev' into stable

peace-maker commented on pwnlib/tubes/ssh.py:
```diff
@@ -903,6 +920,9 @@ def process(self, argv=None, executable=None, tty=True, cwd=None, env=None, time
             self.error("executable / argv[0] must be a string: %r" % executable)
         executable = bytearray(packing._need_bytes(executable, min_wrong=0x80))
 
+        # Validate that all key of env do not contain '=' 
```
This comment seems like a left over.

goreil committed:
ssh.py delete leftover comment

goreil committed:
Merge branch 'stable' of github.com:Youheng-Lue/pwntools into stable

peace-maker commented:
Thank you! This'll make e.g. overthewire solve scripts cleaner 🔥
